pkg/operator/apis/monitoring/v1/types.go

```diff
@@ -1406,6 +1418,189 @@ type NodeMonitoring struct {
+	relabelCfgs, err := relabelingsForSelector(nm.Spec.Selector, nm)
+	if err != nil {
+		return nil, err
+	}
+
+	relabelCfgs = append(relabelCfgs,
+		&relabel.Config{
+			Action:      relabel.Replace,
+			Replacement: nm.Name,
+			TargetLabel: "job",
+		},
+		&relabel.Config{
+			Action:       relabel.Replace,
+			SourceLabels: prommodel.LabelNames{"__meta_kubernetes_node_name"},
+			TargetLabel:  "node",
+		},
+		&relabel.Config{
+			Action:       relabel.Replace,
+			SourceLabels: prommodel.LabelNames{"__meta_kubernetes_node_name"},
+			Replacement:  `$1:metrics`,
+			TargetLabel:  "instance",
+		},
+		// Force target labels so they cannot be overwritten by metric labels.
+		&relabel.Config{
+			Action:      relabel.Replace,
+			TargetLabel: "project_id",
+			Replacement: projectID,
+		},
+		&relabel.Config{
+			Action:      relabel.Replace,
+			TargetLabel: "location",
+			Replacement: location,
+		},
+		&relabel.Config{
+			Action:      relabel.Replace,
+			TargetLabel: "cluster",
+			Replacement: cluster,
+		},
+	)
+
+	discoveryCfgs := discovery.Configs{
+		&discoverykube.SDConfig{
+			HTTPClientConfig: config.DefaultHTTPClientConfig,
+			Role:             discoverykube.RoleNode,
+			// Drop all potential targets not the same node as the collector. The $(NODE_NAME) variable
+			// is interpolated by the config reloader sidecar before the config reaches the Prometheus collector.
+			// Doing it through selectors rather than relabeling should substantially reduce the client and
+			// server side load.
+			Selectors: []discoverykube.SelectorConfig{
+				{
+					Role:  discoverykube.RoleNode,
+					Field: fmt.Sprintf("metadata.name=$(%s)", EnvVarNodeName),
+				},
+			},
+		},
+	}
+
+	ep := nm.Spec.Endpoints[index]
+	// TODO(macxamin) Add support for int port. String ports are not supported by the Kubernetes spec: https://pkg.go.dev/k8s.io/api/core/v1#NodeDaemonEndpoints.
+	if ep.Port.StrVal != "" || ep.Port.IntVal != 0 {
+		return nil, errors.New("port not supported")
+	}
+	// TODO(macxamin) Add support for HTTPClientConfig. Add it for the user for now.
+	if ep.HTTPClientConfig.BasicAuth != nil || ep.HTTPClientConfig.ProxyConfig.ProxyURL != "" || ep.HTTPClientConfig.TLS != nil {
+		return nil, errors.New("http_client_config not supported")
+	}
+	httpCfg := config.HTTPClientConfig{
+		Authorization: &config.Authorization{
+			CredentialsFile: "/var/run/secrets/kubernetes.io/serviceaccount/token",
+		},
+		TLSConfig: config.TLSConfig{
+			CAFile: "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt",
+		},
+	}
+	return extractScrapeParams(nm.GetKey(), ep, httpCfg, relabelCfgs, discoveryCfgs, nm.Spec.Limits)
+}
+
+// extractScrapeParams extracts the scrape parameters of a NodeMonitoring ScrapeEndpoint.
+func extractScrapeParams(id string, ep ScrapeEndpoint, httpCfg config.HTTPClientConfig, relabelCfgs []*relabel.Config, discoveryCfgs discovery.Configs, limits *ScrapeLimits) (*promconfig.ScrapeConfig, error) {
```

Comment:
hm... something could be improved here. Is it specific to NodeMonitoring (as comment suggests?). I don't think it is at the moment.

Also the function name is pretty generic here.. maybe simply `newPromScrapeConfig` is better? Or `endpointScrapeConfigForNM`, or `func (nm *NodeMonitoring) newPromScrapeConf`? Essentially it feels you tried to mimic https://github.com/GoogleCloudPlatform/prometheus-engine/blob/69d276c973b8e4f00531a3366e4140998a10da4d/pkg/operator/apis/monitoring/v1/types.go#L616 function, but this function is only only used once. 

In this state you could reuse `extractScrapeParam` function in https://github.com/GoogleCloudPlatform/prometheus-engine/blob/69d276c973b8e4f00531a3366e4140998a10da4d/pkg/operator/apis/monitoring/v1/types.go#L616, but not sure if want reuse ScrapeEndpoint for NodeMonitoring, what's the plan there, given my comments above?

Reply:
done. PTAL.